fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneSpellcheckRecordCursor.java

```diff
@@ -0,0 +1,234 @@
+import com.apple.foundationdb.record.ScanProperties;
+import com.apple.foundationdb.record.cursors.BaseCursor;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainerState;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.foundationdb.util.LogMessageKeys;
+import com.google.protobuf.ByteString;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.spell.DirectSpellChecker;
+import org.apache.lucene.search.spell.SuggestWord;
+//import org.slf4j.Logger;
+//import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.apple.foundationdb.record.lucene.DirectoryCommitCheckAsync.getOrCreateDirectoryCommitCheckAsync;
+import static com.apple.foundationdb.record.lucene.IndexWriterCommitCheckAsync.getIndexWriterCommitCheckAsync;
+import static java.util.Comparator.comparing;
+
+public class LuceneSpellcheckRecordCursor implements BaseCursor<IndexEntry> {
+
+    // TODO: log some stuff.
+    // private static final Logger LOGGER = LoggerFactory.getLogger(LuceneSpellcheckRecordCursor.class);
+    @Nonnull
+    private final Executor executor;
+    @Nonnull
+    private final IndexMaintainerState state;
+    private final int limit;
+    @Nonnull
+    private final String wordToSpellCheck;
+    @Nonnull
+    private final DirectSpellChecker spellchecker;
+    @Nullable
+    private final FDBStoreTimer timer;
+
+    @Nullable
+    private List<IndexEntry> spellcheckSuggestions = null;
+    private int currentPosition = 0;
+    @Nullable
+    private final Tuple groupingKey;
+    private final List<String> fields;
+
+
+    public LuceneSpellcheckRecordCursor(@Nonnull final String value,
+                                        @Nonnull final Executor executor,
+                                        final ScanProperties scanProperties,
+                                        @Nonnull final IndexMaintainerState state,
+                                        @Nullable Tuple groupingKey,
+                                        final String[] fieldNames) {
+        if (value.contains(":")) {
+            String[] fieldAndWord = value.split(":", 2);
+            // only check the validity of the field if the documents are not grouped.
+            if (groupingKey != null
+                    && groupingKey.size() == 0
+                    && Arrays.stream(fieldNames).noneMatch(name -> name.equals(fieldAndWord[0]))) {
+                throw new RecordCoreException("Invalid field name in Lucene index query")
```

Comment:
As I understand whether the index has a grouping key does not necessarily indicate whether the Lucene field name is a concatenated one. A nested field could also lead to a concatenated field name but it might not have grouping key right?

Reply:
And that concatenation should be reflected in `fieldNames`, against which it is compared, right?